org.jcryptool.visual.rabin/src/org/jcryptool/visual/rabin/HandleCryptosystemTextbook.java

```diff
@@ -0,0 +1,974 @@
+			return;
+		}
+
+		String pattern = "[0-9a-fA-F]+"; 
+		String ciphertext = cstb.textSelector.getText().getText();
+		ciphertext = ciphertext.replaceAll("\\s+", ""); 
+		if(!ciphertext.matches(pattern)) {
+			String strOnlyHexAllowed = Messages.HandleFirstTab_strOnlyHexAllowed;
+			cstb.txtDecryptWarning.setText(strOnlyHexAllowed);
+			showControl(cstb.txtDecryptWarning);
+			if(!cstb.txtCiphertextSegments.getText().isEmpty()) {
+				Display.getDefault().timerExec(3000, new Runnable() {
+					
+					@Override
+					public void run() {
+						hideControl(cstb.txtDecryptWarning);
+					}
+				});
+				
+			}
+			
+			return;
+		}
+		
+		boolean checkLength = ciphertext.length() % guiHandler.blocklength == 0;
+		
+		if(!checkLength) {
+			String strLengthCipher = Messages.HandleFirstTab_strLengthCipher;
+			cstb.txtDecryptWarning.setText(MessageFormat.format(
+					strLengthCipher,
+					(guiHandler.blocklength / 2)));
+			showControl(cstb.txtDecryptWarning);
+			return;
+		}
+		
+		hideControl(cstb.txtDecryptWarning);
+		
+		ciphertextList = guiHandler.rabinFirst.parseString(ciphertext, guiHandler.blocklength);
+		try {
+			plaintexts = guiHandler.rabinFirst.getAllPlaintextsFromListOfCiphertextblocks(ciphertextList, guiHandler.radix);
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			cstb.txtEncryptionWarning.setText("Attention: the computation takes too much time. Please generate a greater N.");
+			guiHandler.showControl(cstb.txtEncryptionWarning);
+
+		}
+		initializeClickedPlaintexts(plaintexts.size(), possiblePlaintextsSize);
+		String ciphertextblocksWithSeparator = guiHandler.rabinFirst.getStringWithSepForm(ciphertextList, guiHandler.separator);
+		cstb.txtCiphertextSegments.setText(ciphertextblocksWithSeparator);
+		
+		setChooseBlock(cstb.cmbChooseBlock, plaintexts.size());
+		cstb.cmbChooseBlock.select(0);
+		
+		
+		int idx = cstb.cmbChooseBlock.getSelectionIndex();
+		int elem = idx + 1;
+		markCiphertext(cstb.txtCiphertextSegments, elem);
+			
+		setPossiblePlaintexts(idx, cstb.txtFirstPlaintext, cstb.txtSecondPlaintext, 
+				cstb.txtThirdPlaintext, cstb.txtFourthPlaintext);
+		
+		updateChosenPlaintexts(idx, cstb, GUIHandler.colorSelectControlBG, GUIHandler.colorDeselectControlBG, GUIHandler.colorSelectControlFG, GUIHandler.colorDeselectControlFG);
+		currentSelectedPlaintexts = new LinkedHashMap<String, String>();
+		cstb.txtChosenPlaintexts.setText("");
+	}
+		
+	
+	
+	private void resetChosenPlaintexts(int idx, CryptosystemTextbookComposite cstb) {
+		updateChosenPlaintexts(idx, cstb, GUIHandler.colorSelectControlBG, GUIHandler.colorDeselectControlBG, GUIHandler.colorSelectControlFG, GUIHandler.colorDeselectControlFG);
+		cstb.txtChosenPlaintexts.setText("");
+	}
+	
+	
+	
+	private String getCurrentSelectedPlaintexts() {
+		String ret = "";
+		Set<String> keys = currentSelectedPlaintexts.keySet();
+		for(String key : keys) {
+			ret += currentSelectedPlaintexts.get(key);
+		}
+		return ret;
```

Comment:
Can be written as nice one-liner with Java [streams](https://stackify.com/streams-guide-java-8/)
```suggestion
		return currentSelectedPlaintexts.values().stream().collect(Collectors.joining());
```